File: AlgorithmsProject/src/in/algorithms/sort/MergeSort.scala
package in.algorithms.sort

object MergeSort {
  def mergesort[T](xs: List[T])(lt: (T, T) => Boolean): List[T] = {

    val n = xs.length / 2
    if (n == 0) xs
    else {
      def merge(xs: List[T], ys: List[T]): List[T] = {
        (xs, ys) match {
          case (Nil, ys) => ys
          case (xs, Nil) => xs
          case (x :: xs1, y :: ys1) => if (lt(x, y)) x :: merge(xs1, ys) else y :: merge(xs, ys1)
        }

      }
      val (fst, sec) = xs splitAt (n)
      merge(mergesort(fst)(lt), mergesort(sec)(lt))
    }

  }
  def main(args: Array[String]) {
  println(mergesort(List(2, 1, 3, 6, 5, 4))((x,y) => x < y))
  println(mergesort(List("mangoes", "oranges", "apples", "lemons"))((x,y) => x.compareTo(y)<0))
}
}

File: AlgorithmsProject/src/in/algorithms/sort/MergeSort.java
package in.algorithms.sort;

import java.util.ArrayList;
import java.util.List;

public class MergeSort {
    public static <T extends Comparable<T>> List<T> mergesort(List<T> list) {
        if (list == null || list.size() <= 1) return list;
        int mid = list.size() / 2;
        List<T> left = mergesort(list.subList(0, mid));
        List<T> right = mergesort(list.subList(mid, list.size()));
        return merge(left, right);
    }

    private static <T extends Comparable<T>> List<T> merge(List<T> left, List<T> right) {
        List<T> result = new ArrayList<>();
        int i = 0, j = 0;
        while (i < left.size() && j < right.size()) {
            if (left.get(i).compareTo(right.get(j)) <= 0) {
                result.add(left.get(i++));
            } else {
                result.add(right.get(j++));
            }
        }
        while (i < left.size()) result.add(left.get(i++));
        while (j < right.size()) result.add(right.get(j++));
        return result;
    }
}
